sdk/azcore/cloud/cloud.go

```diff
@@ -0,0 +1,192 @@
+var WellKnownClouds = map[Name]Configuration{}
+
+// Name identifies a cloud.
+type Name string
+
+const (
+	// AzureChina is a global constant identifying Azure China.
+	AzureChina Name = "AzureChina"
+	// AzureGovernment is a global constant identifying Azure Government.
+	AzureGovernment Name = "AzureGovernment"
+	// AzurePublicCloud is a global constant identifying Azure Public Cloud.
+	AzurePublicCloud Name = "AzurePublicCloud"
+)
+
+func init() {
+	clouds, _ := getConfigurationsFromMetadata(armMetadata)
+	for _, cloud := range clouds {
+		switch cloud.Name {
+		case "AzureChinaCloud":
+			WellKnownClouds[AzureChina] = cloud
+		case "AzureCloud":
+			WellKnownClouds[AzurePublicCloud] = cloud
+		case "AzureUSGovernment":
+			WellKnownClouds[AzureGovernment] = cloud
+		}
+	}
+}
+
+// ServiceName identifies a cloud service.
+type ServiceName string
+
+// ResourceManager is a global constant identifying Azure Resource Manager.
+const ResourceManager ServiceName = "resourceManager"
+
+// ServiceConfiguration configures a specific cloud service such as Azure Resource Manager.
+type ServiceConfiguration struct {
+	// Audiences determine the authentication scope used to authorize access to the service.
+	Audiences []string
+	// Endpoint is the service's URL.
+	Endpoint string
+}
+
+// Configuration configures a cloud.
+type Configuration struct {
+	// LoginEndpoint is the URL of the cloud's Azure Active Directory.
+	LoginEndpoint string
+	// Name is the name of the cloud.
+	Name string
+	// Services contains configuration for the cloud's services.
+	Services map[ServiceName]ServiceConfiguration
+}
+
+// getConfigurationsFromMetadata unmarshals Configuration objects from ARM endpoint metadata
+func getConfigurationsFromMetadata(b []byte) ([]Configuration, error) {
+	var raw []cloudConfiguration
+	if err := json.Unmarshal(b, &raw); err != nil {
+		return nil, err
+	}
+
+	ret := make([]Configuration, 0, len(raw))
+	for _, r := range raw {
+		c := Configuration{
+			LoginEndpoint: r.Authentication.LoginEndpoint,
+			Name:          r.Name,
+			Services: map[ServiceName]ServiceConfiguration{
+				ResourceManager: {
+					Audiences: r.Authentication.Audiences, Endpoint: r.ResourceManager,
+				},
+			},
+		}
+		ret = append(ret, c)
+	}
+
+	return ret, nil
+}
+
+// types for unmarshaling ARM metadata
+type authentication struct {
+	Audiences        []string `json:"audiences"`
+	IdentityProvider string   `json:"identityProvider,omitempty"`
+	LoginEndpoint    string   `json:"loginEndpoint"`
+	Tenant           string   `json:"tenant,omitempty"`
+}
+
+type cloudConfiguration struct {
+	Authentication  authentication    `json:"authentication,omitempty"`
+	Name            string            `json:"name,omitempty"`
+	ResourceManager string            `json:"resourceManager,omitempty"`
+	Suffixes        map[string]string `json:"suffixes,omitempty"`
+}
+
+// https://management.azure.com/metadata/endpoints?api-version=2019-05-01 (minus Azure Germany)
```

Comment:
Why is Azure Germany left out?

Reply:
It closed in October: https://www.microsoft.com/en-us/cloud-platform/germany-cloud-regions